src/vcpkg/abi.cpp

```diff
@@ -0,0 +1,462 @@
+                                                    current_action->spec,
+                                                    dependency_spec));
+                }
+
+                dependency_abis.emplace_back(dependency_spec.name(), status_it->get()->package.abi);
+            }
+            else
+            {
+                // dependency found in action plan
+                dependency_abis.emplace_back(dependency_spec.name(), dependency_it->public_abi());
+            }
+        }
+        return dependency_abis;
+    }
+
+    void compute_all_abis_ex(const VcpkgPaths& paths,
+                             ActionPlan& action_plan,
+                             const CMakeVars::CMakeVarProvider& var_provider,
+                             const StatusParagraphs& status_db)
+    {
+        const AsyncLazy<std::vector<AbiEntry>> cmake_script_hashes(
+            [&paths]() { get_cmake_script_hashes(paths.get_filesystem(), paths.scripts); });
+        // 1. system abi (ports.cmake/ PS version/ CMake version)
+        const auto common_abi = get_common_abi(paths);
+
+        std::vector<AbiEntry> dependency_abis;
+
+        for (auto it = action_plan.install_actions.begin(); it != action_plan.install_actions.end(); ++it)
+        {
+            auto& action = *it;
+
+            if (!Util::Enum::to_bool(action.build_options.only_downloads))
+            {
+                dependency_abis = get_dependency_abis(action_plan.install_actions.begin(), it, status_db);
+            }
+            auto pre_build_info = std::make_unique<PreBuildInfo>(
+                paths, action.spec.triplet(), var_provider.get_tag_vars(action.spec).value_or_exit(VCPKG_LINE_INFO));
+
+            populate_abi_tag_ex(paths, action, std::move(pre_build_info), dependency_abis, cmake_script_hashes);
+            dependency_abis.clear();
+        }
+
+        // get prebuildinfo (not in parallel)
+        // populate abi tag
+    }
+
+    void compute_all_abis(const VcpkgPaths& paths,
+                          ActionPlan& action_plan,
+                          const CMakeVars::CMakeVarProvider& var_provider,
+                          const StatusParagraphs& status_db)
+    {
+        Cache<Path, Optional<std::string>> grdk_cache;
+
+        std::vector<std::reference_wrapper<InstallPlanAction>> must_make_private_abi;
+
+        for (auto& action : action_plan.install_actions)
+        {
+            if (action.abi_info.has_value()) continue;
+
+            must_make_private_abi.push_back(action);
+        }
+
+        std::vector<Optional<std::vector<AbiEntry>>> action_private_abis(must_make_private_abi.size());
+
+        // get "private" abi
+        parallel_transform(must_make_private_abi.begin(),
+                           must_make_private_abi.size(),
+                           action_private_abis.begin(),
+                           [&](auto&& action) {
+                               return populate_abi_tag(paths, action, grdk_cache);
```

Comment:
This is super not thread safe.

I'm not going to continue reviewing further at this point; anything you're going to parallelize needs to come with strong proof that it's safe to do.

Reply:
I know. I'll remove this.